### ISSUE TITLE
"idle" electron periodically pegs the cpu on a single core vm

### ISSUE THREAD

**Issue by bughit** [electron/electron#14319]
* Output of `node_modules/.bin/electron --version`: v3.0.0-beta.7 x86
* Operating System (Platform and Version): win7 sp1 (x86 & x64) single core

**Expected Behavior**
no cpu pegging when electron is idle

**Actual behavior**
cpu is is periodically pegged
![image](https://user-images.githubusercontent.com/895588/44633561-1426b800-a95b-11e8-9f65-35f98f4c00e0.png)

**To Reproduce**
This is triggered by simply creating winston.transports.File.  app.disableHardwareAcceleration contributes as well.  Single core is required to repro.  I first discovered it on a single core x86 win7 vm, could not repro on a 4 core x64 vm, but lowering the cores to 1, reproduced.

AFAICT from some profiling, no js is running, electron should be idle.  Does not happen in node.

run he following app (npm install first)

##### package.json
```json
{
  "dependencies": {
    "winston": "^3.0.0"
  }
}
```

##### index.js
```js
'use strict';

const winston = require('winston'),
      {app} = require('electron');

app.disableHardwareAcceleration();

const winstonTransports = [
  new winston.transports.File({
    filename: 'test.log',
    maxsize: 10 * 2**20,
    maxFiles: 2,
    tailable: true,
  }),
];
```

sometimes the initial spike is brief and you have to wait for the next

**Comment by bughit** [comment]
repro can be simplified, winston can be replaced by simply doing async file io

```js
'use strict';

require('electron').app.disableHardwareAcceleration();

const fs = require('fs'),
      fd = fs.openSync('test.log', 'a');


log();
setInterval(log, 250);

function log() {
  fs.appendFile(fd, 'log line\n', error => {});
}
```

on one single core vm this pegs the cpu continuously, on another periodically.  I thought disableHardwareAcceleration makes this worse (easier to repro), not sure, leaving it in, just in case.

@zcbenz Node is ok, so it seems like a serious electron bug.

**Comment by bughit** [comment]
@zcbenz Could you please give some guidance on this. Our code would be running on customer vms and some of them are single core, this is a show stopper for us.  I also found that sync or async stdin reading triggers this ( `fs.readFileSync(0, 'utf8')`)

I would be happy to fix this but don't know enough about libuv internals and it's interaction with electron.  Is there any chance of this addressed?  Could you comment on the proposed solution/workaround https://github.com/electron/electron/issues/11699#issuecomment-368581866

**Comment by bughit** [comment]
@zcbenz @deepak1556 

I can appreciate that single-core is not your highest priority, since it basically does not exist any more on physical hardware, but if you are not going to address this, you should make dual core a documented minimum requirement.

**Comment by bughit** [comment]
@zcbenz

I tried @gshively11's fix from #12026, changing `NumberOfConcurrentThreads` to 2

```
loop->iocp = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 2); // last arg changed from 1 to 2
```

Seems to resove this.  What do you think of this approach?  Would you accept a pull-request that passes 2 on windows when NUMBER_OF_PROCESSORS=1

**Comment by zcbenz** [comment]
@bughit Thanks for looking into it, the solution in #12026 looks good to me.

Before creating a PR in electron, can you create a PR at https://github.com/libuv/libuv first? I would like to know how libuv's developers think.

**Comment by zcbenz** [comment]
@bughit I'm good having the patch in Electron if there is no bad side effect, I just want to know if libuv developers can find problems of the patch.

**Comment by gshively11** [comment]
@bughit thanks for the shoutouts.  We're no longer impacted by this bug, but I hope you can drive a fix home.

**Comment by bughit** [comment]
> Did anyone have any insight on why this happens in electron and not in node?

[another](https://github.com/electron/electron/issues/11699#issuecomment-368581866) proposed fix involved [`NodeBindingsWin::PollEvents`](https://github.com/electron/electron/blob/master/atom/common/node_bindings_win.cc#L18) which is specific to electron.  [Analysis](https://github.com/electron/electron/issues/11699#issuecomment-365972258) from there:

> After investigation of the source code of Electron, it was turned out that the problem was caused by atom::NodeBindingsWin::PollEvents method. The call of ::GetQueuedCompletionStatus on thread that differs from libuv poll leaded to errors inside uv_run API. After replacing the pair of IOCP functions to polling on uv_loop_alive, the problem had been sorted out.

**Comment by zcbenz** [comment]
@refack Electron does not run a libuv loop directly since the main message loop has to be a GUI message loop. Instead Electron listens to the backend fd of Node's loop in a new thread, and iterates the libuv loop when an event happens.

```c++
// Following code is running in a new thread.
void NodeBindingsWin::PollEvents() {
  // If there are other kinds of events pending, uv_backend_timeout will
  // instruct us not to wait.
  DWORD bytes, timeout;
  ULONG_PTR key;
  OVERLAPPED* overlapped;

  timeout = uv_backend_timeout(uv_loop_);

  // This call is blocking until something happens in the uv loop.
  GetQueuedCompletionStatus(uv_loop_->iocp, &bytes, &key, &overlapped, timeout);

  // Give the event back so libuv can deal with it.
  if (overlapped != NULL)
    PostQueuedCompletionStatus(uv_loop_->iocp, bytes, key, overlapped);
}
```

**Comment by refack** [comment]
Thank you @bughit and @zcbenz... 
I'm trying to repro on a single CPU VM (running on my 2x2 CPU even with limited affinity doesn't work, I have a feeling this is related to the other places that call `CreateIoCompletionPort` with `NumberOfConcurrentThreads == 0`).

So I'm guessing what happens is that there's a uv thread that's blocking on this `iocp` and that makes the Electron thread run in a busy loop.

Meanwhile I'm thinking if you are already "breaking the abstraction" and anyway using `uv_loop_->iocp` directly, you could test replacing it with one that has `NumberOfConcurrentThreads == 2` in:
https://github.com/electron/electron/blob/47bf8e1bb3b38e7eae0d57873cb1fce874195317/atom/common/node_bindings.cc#L159-L167

**Comment by bughit** [comment]
something like this?

```c++
NodeBindingsWin::NodeBindingsWin(BrowserEnvironment browser_env)
    : NodeBindings(browser_env) {
  char *numProcessors = getenv("NUMBER_OF_PROCESSORS");
  if (numProcessors && strcmp(numProcessors, "1") == 0) {
    if (uv_loop_->iocp && uv_loop_->iocp != INVALID_HANDLE_VALUE)
      CloseHandle(uv_loop_->iocp);
    uv_loop_->iocp = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 2);
  }
}
```

**Comment by refack** [comment]
> something like this?

Yes. But this needs to be tested because of the `uv_loop_ = uv_default_loop();` case. If there are any handles still queued up on the loop, I'm not sure what will happen.

In parallel I'm reproducing, and figuring out if libuv can provide an abstracted cross-platform solution (something like a blocking `uv_loop_peek(uv_loop_)`)

**Comment by bughit** [comment]
> Yes. But this needs to be tested because of the uv_loop_ = uv_default_loop(); case.

AFAICT from inspecting the code and debugging the NodeBindings::BROWSER case, the first call to uv_default_loop() is from the NodeBindings constructor. So does that mean it should be fine, since you are replacing iocp right after its initial creation?

**Comment by refack** [comment]
> So does that mean it should be fine, since you are replacing iocp right after its initial creation?

Sound highly likely to be safe. Maybe add a `CHECK` just to make sure...

**Comment by refack** [comment]
You could:
```cpp
CHECK_EQ(0, uv_loop_->active_handles);
CHECK_EQ(0, uv_loop_->active_reqs.count);
```

**Comment by bpasero** [comment]
> I can appreciate that single-core is not your highest priority, since it basically does not exist any more on physical hardware

VSCode is also very interested in a fix, we have users hitting this in VM environments all the time 👍

**Comment by refack** [comment]
@bpasero see https://github.com/electron/electron/pull/15975

BTW: I had a real trouble installing VS2017 on a single core machine, since the VS installer is an electron app 🤦‍♂️

**Comment by bughit** [comment]
fixed by #15975
